src/liballoc/arc.rs

```diff
@@ -352,39 +373,40 @@ impl<T: Clone> Arc<T> {
     /// This is also referred to as a copy-on-write operation because the inner
     /// data is cloned if the reference count is greater than one.
     ///
-    /// This method is marked **unsafe** because it is racy if weak pointers
-    /// are active.
-    ///
     /// # Examples
     ///
     /// ```
     /// # #![feature(arc_unique)]
     /// use std::sync::Arc;
     ///
-    /// # unsafe {
     /// let mut five = Arc::new(5);
     ///
     /// let mut_five = five.make_unique();
-    /// # }
     /// ```
     #[inline]
     #[unstable(feature = "arc_unique")]
-    #[deprecated(since = "1.2.0",
-                 reason = "this function is unsafe with weak pointers")]
-    pub unsafe fn make_unique(&mut self) -> &mut T {
-        // FIXME(#24880) potential race with upgraded weak pointers here
-        //
+    pub fn make_unique(&mut self) -> &mut T {
         // Note that we hold a strong reference, which also counts as a weak
         // reference, so we only clone if there is an additional reference of
         // either kind.
-        if self.inner().strong.load(SeqCst) != 1 ||
-           self.inner().weak.load(SeqCst) != 1 {
+        //
+        // The use of usize::MAX here acts as a "lock" while we try to determine
+        // whether this thread is the unique owner.
+        if self.inner().weak.compare_and_swap(1, usize::MAX, SeqCst) != 1 {
             *self = Arc::new((**self).clone())
+        } else {
+            let need_clone = self.inner().strong.load(SeqCst) != 1;
+            self.inner().weak.store(1, Relaxed); // release lock immediately;
+                                                 // we don't need to hold it while cloning
+            if need_clone { *self = Arc::new((**self).clone()); }
         }
```

Comment:
We should only clone if there are other live strong references; the correct implementation is something like the following (untested, possibly buggy):

```
if self.inner().strong.compare_and_swap(1, 0, Release) != 1 {
    // Another strong pointer exists; clone.
    *self = Arc::new((**self).clone());
} else if self.inner().weak.load(SeqCst) != 1 {
    // There are no strong pointers left, but there are remaining weak
    // pointers; steal the contents.
    // FIXME: Is there a better way to write this?
    let weak = Weak { _ptr: self._ptr };
    self._ptr = 0;
    let stolen_data = ptr::read(&(*weak._ptr).data);
    *self = Arc::new(stolen_data);
} else {
    // No other strong or weak pointers; just revive this Arc.
    self.inner().strong = 1;
}
```
No spinlock necessary.